### PR TITLE
Add benchmarking for parachain runtime `ump` pallet

### PR DESCRIPTION
**Pull request by hirschenberger**
Part of #3850

@KiChjang Can you add me to the tracking list?

### CODE REVIEW

File: runtime/parachains/src/ump.rs
Comment (kianenigma) [inline runtime/parachains/src/ump.rs:216]:
```suggestion
		/// Weight information for extrinsics in this pallet.
```

File: runtime/parachains/src/ump/benchmarking.rs
Comment (kianenigma) [inline runtime/parachains/src/ump/benchmarking.rs:65]: this is basically using the value of the `ActiveConfig` in genesis, not the real on on-chain.

I feel like coding it like this is actually misleading. A better approach would be to hardcode some maximum value for it, and ensuring that when setting it, we don't go over it. See an example here: https://github.com/paritytech/polkadot/pull/3876/files

File: runtime/parachains/src/ump/benchmarking.rs
Comment (kianenigma) [inline runtime/parachains/src/ump/benchmarking.rs:75]: same

**Review by kianenigma — COMMENTED** [review]
a minor note, otherwise LGTM

File: runtime/parachains/src/ump/benchmarking.rs
Comment (hirschenberger) [reply runtime/parachains/src/ump/benchmarking.rs:65]: I see, but how can I determine this upper bound? 
Isn't it a delicate point bounding on-chain values, which in principle stand under the control of democracy.

Comment (kianenigma) [reply runtime/parachains/src/ump/benchmarking.rs:65]: Some reasonable defaults should be enough, and no if some upper bounds exists even the governance should not (easily) be able to set higher values. @pepyakin might be able to chime in here.

Comment (hirschenberger) [reply runtime/parachains/src/ump/benchmarking.rs:65]: I meant it the other way around. Governance  should not be **limited** by bounds set in code by a single developer.

But I'm still a noob and have not so much insight in the delicate gears that make a chain safe and governance truly  democratic. It was just a feeling and an open question.

Comment (kianenigma) [reply runtime/parachains/src/ump/benchmarking.rs:65]: > I meant it the other way around. Governance should not be limited by bounds set in code by a single developer.

The governance is not and will not be bound at the end of the day by any of these because by the same token that they can set this parameter, they can also change the code that contains this parameter.

I agree that the value should not be arbitrary, therefore I mentioned someone more knowledgable about parachains.

File: runtime/parachains/src/mock.rs
Comment (kianenigma) [inline runtime/parachains/src/mock.rs:141]: Please call this `TestWeightInfo` instead of `()`.

File: runtime/parachains/src/ump.rs
Comment (kianenigma) [inline runtime/parachains/src/ump.rs:179]:
```suggestion
impl WeightInfo for TestWeightInfo {
```

File: runtime/parachains/src/ump.rs
Comment (kianenigma) [inline runtime/parachains/src/ump.rs:109]: This benchmark is a bit wrong IMO. `WeightInfo::process_upward_message` is measuring the execution of this entire code path, including the execution of the xcm itself. Yet, you are summing the weight used in the execution of the xcm with this weight. 

Ideally, the xcm operation itself should be something as close as possible to noop in the benchmark (`Transact` with `remark` might actually _be_ good for that) and then you could add them up.

That being said though, this all is kinda fine since you are over-couting the weight, which is less of a big deal.

**Review by kianenigma — COMMENTED** [review]
aside from my other note, needs to benchmark `initializer_on_new_session` as well and return its weight. I will do the same for #3876 as well. Everything else looks good to me.

File: runtime/parachains/src/ump.rs
Comment (hirschenberger) [reply runtime/parachains/src/ump.rs:109]: I do exactly this, the xcm operation is a Transact-remark with different sizes, to take the decoding effort into account but having a minimal xcm-execution overhead. https://github.com/paritytech/polkadot/blob/f88e7ee92e9c915139d70332c4e4f7e199b3499b/runtime/parachains/src/ump/benchmarking.rs#L55-L59

I can also just benchmark the xcm-decoding https://github.com/paritytech/polkadot/blob/f88e7ee92e9c915139d70332c4e4f7e199b3499b/runtime/parachains/src/ump.rs#L102

Or is there a even a benchmarked weight for this somewhere? If not it may make sense to add one and take decoding time into account in all places where it's used.

File: runtime/parachains/src/ump.rs
Comment (hirschenberger) [reply runtime/parachains/src/ump.rs:377]: I'm using a FP-language in my fulltime job, so this seems more natural to me than a loop. But I'm not religious about that.

File: runtime/parachains/src/ump.rs
Comment (hirschenberger) [reply runtime/parachains/src/ump.rs:800]: You're right, this is a relic from an older version.

File: runtime/parachains/src/ump.rs
Comment (kianenigma) [inline runtime/parachains/src/ump.rs:173]:
```suggestion
/// Weight information of this pallet. 
pub trait WeightInfo {
```

File: runtime/parachains/src/ump.rs
Comment (kianenigma) [inline runtime/parachains/src/ump.rs:179]:
```suggestion
/// fallback implementation
```

File: runtime/parachains/src/ump.rs
Comment (kianenigma) [reply runtime/parachains/src/ump.rs:109]: Yeah this is really tricky. I am fine with it given

1. This is an overestimate and is therefore safe.
2. please make `// NOTE: ` and explain what is being done, and why it is not accurate.

Comment (kianenigma) [reply runtime/parachains/src/ump.rs:109]: ideally, I would break down the function in ways that it can be easily weighed, but a story for another day.

File: runtime/parachains/src/ump/benchmarking.rs
Comment (kianenigma) [reply runtime/parachains/src/ump/benchmarking.rs:57]: This is not fixed.

Comment (hirschenberger) [reply runtime/parachains/src/ump/benchmarking.rs:62]: I need a hint for sensible bound values.

Comment (hirschenberger) [reply runtime/parachains/src/ump/benchmarking.rs:62]: @kianenigma @pepyakin Can someone give me a hint for reasonable default values? I really like to get this PR landed before christmas 😄

Comment (pepyakin) [reply runtime/parachains/src/ump/benchmarking.rs:62]: Sorry.

Currently on kusama and polkadot this value is set to 50 KiB. I think this is a sensible value. I think it may be possible we decrease this value even. (Decreasing would mean that the value would go out of sync with the benchmarking but that's OK as long as it goes into the conservative direction)

Comment (hirschenberger) [reply runtime/parachains/src/ump/benchmarking.rs]: @kianenigma I added a max-bound as suggested with the value @pepyakin proposed. Can I have another review please?

File: runtime/parachains/src/configuration.rs
Comment (pepyakin) [inline runtime/parachains/src/configuration.rs:792]: Perhaps, this should also be reflected in `check_consistency`?

Comment (hirschenberger) [reply runtime/parachains/src/configuration.rs:792]: @pepyakin As I see it, I have to change the generic parameter of `HostConfiguration` from `Config::BlockNumber` to `Config` make get access to `MaxUpwardMessageSizeBound`. Is this ok?

Comment (hirschenberger) [reply runtime/parachains/src/configuration.rs:792]: Hey @pepyakin When I moved the checks to `check_consistency`, running any chain fails because in `node/service/src/chain_spec.rs` and `bridges/bin/rialto/node/src/chain_spec.rs` the max-size is set to `1024*1024` and that is way higher than the 50KiB you proposed. What do you suggest to solve this?

Comment (pepyakin) [reply runtime/parachains/src/configuration.rs:792]: Ah, good to hear that `check_consistency` gets its job done. I say we decrease those numbers to 50 KiB.

Comment (hirschenberger) [reply runtime/parachains/src/configuration.rs:792]: Hey @pepyakin or @kianenigma I'm stuck here with an issue and maybe you can help me. ATM the `service_overweight` bench fails bc. I'm unable to create an overweight message. When I limit the message size to 50KiB I won't reach the current weight limit to make it overweight.

Ideally one can create an arbitrary-weighted call by using `fill_block( req_weight / max_weight )` I tried it but it seems the execution of `fill_block` in `XcmExecutor::execute_xcm` does not consume any weight at all and so will not issue a `XcmError::WeightLimitReached`.

My idea was to create a `pallet_utilities::batch` with as many `remark` calls to reach overweight, but I ran into compiler issues with `T::Config` vs. `pallet_utilities::Config`. Do you have a better idea how to create a heavyweight call?
If the later approach is ok to you, I'd try to fix the issues (or ask more specific questions).

File: bridges/bin/rialto/node/src/chain_spec.rs
Comment (emostov) [inline bridges/bin/rialto/node/src/chain_spec.rs:268]: Is there any specific rational for this change?

Comment (hirschenberger) [reply bridges/bin/rialto/node/src/chain_spec.rs:268]: @pepyakin Proposed that value here: https://github.com/paritytech/polkadot/pull/3889#discussion_r758584658

File: runtime/parachains/src/ump/benchmarking.rs
Comment (emostov) [inline runtime/parachains/src/ump/benchmarking.rs:75]: Is this getting used anywhere?

**Review by emostov — APPROVED** [review]
This makes sense to me. Would be good if @KiChjang , @pepyakin , or @kianenigma could take another look

File: runtime/parachains/src/ump/benchmarking.rs
Comment (emostov) [inline runtime/parachains/src/ump/benchmarking.rs:81]: I don't think its a big deal, but would it make more sense to name this benchmark `clean_ump_after_outgoing`?

Comment (hirschenberger) [reply runtime/parachains/src/ump/benchmarking.rs]: This is indeed confusing, as I first benchmarked `perform_outgoing_para_cleanup` and switched it later to only bench it's inner function `clean_ump_after_outgoing`, as the weight used by this function is summed up and returned in `perform_outgoing_para_cleanup`.

File: runtime/parachains/src/ump.rs
Comment (kianenigma) [inline runtime/parachains/src/ump.rs:122]: bad name -- should be called `maybe_msg_and_weight` or something of that sort.

File: runtime/parachains/src/ump.rs
Comment (drahnr) [inline runtime/parachains/src/ump.rs:765]: the `mod tests` has been made a separate file, so moving the changeset to `runtime/parachains/src/ump/tests.rs` should resolve the conflict.

**Review by drahnr — COMMENTED** [review]
Generally LGTM, I defer the final review to @pepyakin who knows the intricacies and possible side effects better, and hence can judge if the benchmarks measure what the ought to measure :)

**Review by pepyakin — APPROVED** [review]
Looks good. I am OK with merging it as is, possibly following up with nit fixes. Thanks for your work and I am sorry about it took so much time to get the review.

File: runtime/parachains/src/ump.rs
Comment (hirschenberger) [reply runtime/parachains/src/ump.rs:398]: When I do `pub(in crate::ump::benchmarking)` the compiler complains: `error[E0742]: visibilities can only be restricted to ancestor modules`

**Comment by drahnr** [comment]
Let's merge this after 0.9.17 is cut.

File: runtime/parachains/src/ump.rs
Comment (hirschenberger) [reply runtime/parachains/src/ump.rs]: Yes, `Weight::MAX` seems a better value

File: runtime/parachains/src/ump/benchmarking.rs
Comment (hirschenberger) [reply runtime/parachains/src/ump/benchmarking.rs]: I changed the function name to indicate that the created message will have at least `size` bytes.

I also changed the code to create an empty message (without remark payload) to determine the encoded size and use this to create the actual message as close as possible (but always bigger) to `size`.  Ensuring that with an assertion.

**Comment by emostov** [comment]
> Let's merge this after 0.9.17 is cut.

@drahnr is this ok to merge now?

**Comment by drahnr** [comment]
Yes, feel free to merge

**Comment by kianenigma** [comment]
bot merge

**Comment by hirschenberger** [comment]
> @hirschenberger can you put your DOT or KSM address in your PR description? Then we can propose a tip for you!

Oh nice, thank you. I'm glad I got this PR finally landed. My DOT address is:
`1ZApdbwrNnT1Hey4xeTAH6SZPDgbA9ZVftNuu9jTRbkc2Bg`

**Pull request merged**
